File: smarttrade-web/src/features/chart-analysis/domain/entities/chartAnnotation.ts
/** Whether a {@link SupportResistanceZone} is acting as a floor or a ceiling. */
export type ZoneKind = 'support' | 'resistance'

interface BaseAnnotation {
  id: string
  instrumentSymbol: string
  createdAt: Date
  label?: string
  note?: string
}

/** A horizontal price band the user marked as support or resistance. */
export interface SupportResistanceZone extends BaseAnnotation {
  type: 'support_resistance'
  kind: ZoneKind
  priceHigh: number
  priceLow: number
}

/**
 * Two parallel trendlines (upper/lower bound) drawn between two points
 * in time, used as a reference for the prevailing trend structure.
 */
export interface ReferenceChannel extends BaseAnnotation {
  type: 'reference_channel'
  anchorTime: Date
  endTime: Date
  upperPriceStart: number
  lowerPriceStart: number
  upperPriceEnd: number
  lowerPriceEnd: number
}

/**
 * A projected expansion of an existing {@link ReferenceChannel} (e.g. a
 * Fibonacci-style multiple of the channel width projected forward in
 * time), kept as its own annotation so the original channel is never
 * mutated once drawn.
 */
export interface ChannelExpansion extends BaseAnnotation {
  type: 'channel_expansion'
  /** `ReferenceChannel.id` this expansion was projected from. */
  sourceChannelId: string
  /** Multiple of the source channel's width (e.g. 1.618). */
  expansionFactor: number
  projectedTime: Date
  projectedUpperPrice: number
  projectedLowerPrice: number
}

/**
 * A temporal marker used to track recurring market cycles (e.g. "this
 * looks like the start of a new accumulation cycle").
 */
export interface CycleAnnotation extends BaseAnnotation {
  type: 'cycle'
  cycleStart: Date
  cycleEnd?: Date
  expectedCycleLengthDays?: number
}

/**
 * A saved study mark on an instrument's chart.
 *
 * A discriminated union (on `type`) over the four study types the MVP
 * needs to persist: price zones, reference channels, channel
 * expansions, and cycle-tracking time markers. Every concrete variant is
 * stored in the same IndexedDB `chartAnnotations` table (see
 * `core/persistence/db.ts`) since Dexie is schema-less per record, so
 * adding a fifth study type later doesn't require a migration.
 */
export type ChartAnnotation =
  | SupportResistanceZone
  | ReferenceChannel
  | ChannelExpansion
  | CycleAnnotation

File: smarttrade_app/lib/features/chart_analysis/domain/entities/chart_annotation.dart
import 'package:equatable/equatable.dart';

/// Whether a [SupportResistanceZone] is acting as a floor or a ceiling.
enum ZoneKind { support, resistance }

/// A saved study mark on an instrument's chart.
///
/// Sealed over the four study types the MVP needs to persist: price
/// zones, reference channels, channel expansions, and cycle-tracking
/// time markers. Every concrete subtype is backed by the same
/// `chart_annotations` sqlite table (see `data/models/chart_annotation_model.dart`
/// for the mapping), so adding a new study type later is a data-layer
/// change, not a schema migration.
sealed class ChartAnnotation extends Equatable {
  const ChartAnnotation({
    required this.id,
    required this.instrumentSymbol,
    required this.createdAt,
    this.label,
    this.note,
  });

  final String id;
  final String instrumentSymbol;
  final DateTime createdAt;
  final String? label;
  final String? note;
}

/// A horizontal price band the user marked as support or resistance.
class SupportResistanceZone extends ChartAnnotation {
  const SupportResistanceZone({
    required super.id,
    required super.instrumentSymbol,
    required super.createdAt,
    super.label,
    super.note,
    required this.kind,
    required this.priceHigh,
    required this.priceLow,
  });

  final ZoneKind kind;
  final double priceHigh;
  final double priceLow;

  @override
  List<Object?> get props =>
      [id, instrumentSymbol, createdAt, label, note, kind, priceHigh, priceLow];
}

/// Two parallel trendlines (upper/lower bound) drawn between two points
/// in time, used as a reference for the prevailing trend structure.
class ReferenceChannel extends ChartAnnotation {
  const ReferenceChannel({
    required super.id,
    required super.instrumentSymbol,
    required super.createdAt,
    super.label,
    super.note,
    required this.anchorTime,
    required this.endTime,
    required this.upperPriceStart,
    required this.lowerPriceStart,
    required this.upperPriceEnd,
    required this.lowerPriceEnd,
  });

  final DateTime anchorTime;
  final DateTime endTime;
  final double upperPriceStart;
  final double lowerPriceStart;
  final double upperPriceEnd;
  final double lowerPriceEnd;

  @override
  List<Object?> get props => [
        id,
        instrumentSymbol,
        createdAt,
        label,
        note,
        anchorTime,
        endTime,
        upperPriceStart,
        lowerPriceStart,
        upperPriceEnd,
        lowerPriceEnd,
      ];
}

/// A projected expansion of an existing [ReferenceChannel] (e.g. a
/// Fibonacci-style multiple of the channel width projected forward in
/// time), kept as its own annotation so the original channel is never
/// mutated once drawn.
class ChannelExpansion extends ChartAnnotation {
  const ChannelExpansion({
    required super.id,
    required super.instrumentSymbol,
    required super.createdAt,
    super.label,
    super.note,
    required this.sourceChannelId,
    required this.expansionFactor,
    required this.projectedTime,
    required this.projectedUpperPrice,
    required this.projectedLowerPrice,
  });

  /// [ReferenceChannel.id] this expansion was projected from.
  final String sourceChannelId;

  /// Multiple of the source channel's width (e.g. 1.618).
  final double expansionFactor;
  final DateTime projectedTime;
  final double projectedUpperPrice;
  final double projectedLowerPrice;

  @override
  List<Object?> get props => [
        id,
        instrumentSymbol,
        createdAt,
        label,
        note,
        sourceChannelId,
        expansionFactor,
        projectedTime,
        projectedUpperPrice,
        projectedLowerPrice,
      ];
}

/// A temporal marker used to track recurring market cycles (e.g. "this
/// looks like the start of a new accumulation cycle").
class CycleAnnotation extends ChartAnnotation {
  const CycleAnnotation({
    required super.id,
    required super.instrumentSymbol,
    required super.createdAt,
    super.label,
    super.note,
    required this.cycleStart,
    this.cycleEnd,
    this.expectedCycleLengthDays,
  });

  final DateTime cycleStart;
  final DateTime? cycleEnd;
  final int? expectedCycleLengthDays;

  @override
  List<Object?> get props => [
        id,
        instrumentSymbol,
        createdAt,
        label,
        note,
        cycleStart,
        cycleEnd,
        expectedCycleLengthDays,
      ];
}
